keps/0031-namespace-management.md

```diff
@@ -0,0 +1,167 @@
+
+All namespace creation is managed outside KUDO. Users such as D2iQ would need to find or build automation for the creation of their namespaces. This REQUIRES that KUDO honor the deployment of resources to the "default" namespace, or to the namespace defined by `--namespace` during install.  
+
+KUDO should further require that NO namespaces be allow in manifests for resources deployed during "install".  The concept of defining manifest which land in multiple namespaces is defined in a latter proposal and is a separate concern. Without multi-namespace support, all resources are install to one namespace and namespace must be created prior to installation and that namespace must be specified during installation.
+
+### Proposal 2: Namespace Creation
+
+KUDO should support the creation of a namespace if it does not exist.  Further KUDO will add a configuration for namespace such that at time of namespace creation it is created with the metadata provided. To this end a manifest file for namespace is to be supported in the template folder:
+
+```
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    ca.istio.io/override: "true"
+    istio-injection: enabled
+    katib-metricscollector-injection: enabled
+```
+
+The template supports templates for labels and annotations, however the `name:` metadata is NOT allowed to be set or controlled through templating.   This is to prevent the "templating" of the namespace name such as `name: {.Namespace}-extended`.  KUDO will add the `name:` for metadata when there is a need to create a namespace with metadata using the namespaceManifest file. The `operator.yaml` file is to will be extended to support `namespaceManifest`.  
+
+```
+# operator.yaml
+name: "kubeflow"
+operatorVersion: "0.2.0"
+kudoVersion: "0.10.1"
+kubernetesVersion: 1.15.0
+appVersion: 1.0.0
+namespaceManifest: templates/namespace.yaml
+```
+
+KUDO is to add a `--apply-ns`.  The use of `--namespace` defines the namespace that will be used by the operator.  If the namespace is missing than a failure will occur.  If used in conjunction with `--apply-ns`, than the namespace if missing from the cluster will be created using the manifest file.  If the namespace exists, the metadata will be applied to the existing namespace overwriting the exiting metadata.
+
+The follow rules apply:
+
+1. Missing `namespaceManifest` means that there is no metadata.  Using `--apply-ns` will create namespace without metadata.
+1. `name: {{ .Namespace }}` is not allowed in the manifest file and is considered a failure by the package verifier.  The namespace manifest is only useful to provide metadata to the namespace.  The control over what the name of the namespace is is controlled by KUDO.
+1. No support for static namespace.
+1. If no `namespaceManifest` and no `--namespace` the operator installs to "default".  If used in conjunction with `--apply-ns`, it will be ignored.  There is no overwritting of the "default" namespace.
```

Comment:
What if both `--apply-ns` and `--namespace` are specified but they are different?

This is just an example. To simplify the logic and to make it easier in UX, could we make these two flags "at most specify one of them"? so that we could also avoid the conjunction case

Reply:
UX nits^^

Reply:
I don't think we need another namespace flag: now you need to define rules for all 8 permutations of `--apply-ns`, `--namespace` and `namespaceManifest`. And what are we trying to optimize - the UX of creating a ns without the manifest? I would argue it is not worth it. Adding a `namespaceManifest` to the operator is not a big deal.